# While

Assim como o for, o **while é um laço usado para repetir um comando** um número **finito** de vezes

Enquanto no **for** só conseguimos **fazer o laço em uma lista de valores** (não necessariamente numéricos), no while é possível realizar a execução até que a condição seja satisfeira

**Atenção:**
- Ao usar o while, tenha muito cuidado para o laço não ser infinito!

Se temos uma lista como a mostrada abaixo

In [1]:
lista = ['a',2,'teste','Lucas']

Podemos usar tanto o for quanto o while 
(não vamos mostrar o for com detalhes, voltaremos nesse ponto em outro link)

In [3]:
# Para usar o for, podemos fazer diretamente:
for i in lista:
  print(i)

a
2
teste
Lucas


In [5]:
# Já para o while, devemos percorrer cada elemento dessa lista

# Para isso vamos definir a posição zero
posicao = 0

# E percorrer enquanto a posicao for menor que o tamanho da lista
while posicao < len(lista):  # len(lista) retorna o tamanho dessa lista

  # Para fazer o print, vamos pegar o elemento da lista naquela posição
  print(lista[posicao])

  # E para continuar percorrendo a lista, vamos adicionar +1 na posição atual
  posicao = posicao + 1

a
2
teste
Lucas


### IMPORTANTE

Fazer o **posicao = posicao+1** é muito importante para **evitar que o laço seja inifinito**

Se não fizermos isso, o laço ficará executando pra sempre e poderá quebrar o nosso código

Outra vantagem do while é que podemos usá-lo **mesmo fora de uma lista**

Por exemplo, se queremos executar um processo **ATÉ QUE UMA CONDIÇÃO SEJA SATISFEITA** mas não sabemos quando isso vai acontecer, podemos colocar a condição como aquilo que irá parar o laço.
Para ficar mais claro vamos usar um exemplo, queremos todos os números da sequência de Fibonacci menores que 100.

Obs: a sequência de Fibonacci é dada sempre somando os 2 números anteriores da lista

In [2]:
# Pegando os 2 primeiros valores dessa sequência:
num1 = 1
num2 = 1
num_seguinte = num1 + num2

# Enquanto o número seguinte for menor que 100 eu vou executar
while num_seguinte < 100:
    
    # Quando eu dou um "passo" na sequência o num1 passa a ser o num2 e o num2 passa a ser o num_seguinte
    # E então calculamos novamente o num_seguinte
    # (Uma outra forma de fazer a sequência de Fibonacci e uma explicação melhor sobre ela é mostrada no ex 2)
    # https://github.com/llucaslleall/ProjectEuler/blob/main/Ex_002.ipynb
    
    num1 = num2
    num2 = num_seguinte
    
    # E novamente vamos fazer o mesmo cálculo para encontrar o num_seguinte
    num_seguinte = num1 + num2
    
    # Para conseguir acompanhar vamos printar todos os números
    print(num_seguinte)

3
5
8
13
21
34
55
89
144


Observe que ele printou o 144, mas como se **144 é maior que 100**?

O que acontece é o seguinte:

- Ele achou que **num_seguinte é 55**:
    - Ele faz que num1 = 34 e num2 = 55 -> num_seguinte = 89
        - Ele printa o 89
            - Ele **volta no while** e valida: **89 é menor que 100? Sim! Continua o laço**
            
- Então ele chega em **num_seguinte é 89**:
    - Ele faz que num1 = 55 e num2 = 89 -> num_seguinte = 144
        - Ele printa o 144 **(Observe que aqui ele ainda não validou a condição do while)**
            - Ele **volta no while** e valida: **144 é menor que 100? Não! Quebra o laço**

- Também podemos usar o **break** para forçar a parada de execução, para isso vamos precisar **combinar o while com um if**

- Uma opção com o break é usar o **while True**, que vai executar sempre e **só para através do break (CUIDADO!)**

Fazendo o mesmo exemplo acima com o break:

In [3]:
num1 = 1
num2 = 1
num_seguinte = num1 + num2

while True:
    num1 = num2
    num2 = num_seguinte
    
    num_seguinte = num1 + num2
    
    if(num_seguinte > 100):
        break
    
    print(num_seguinte)

3
5
8
13
21
34
55
89


Nesse caso, como colocamos a condição do break antes do print, ele não printou o valor maior que 100.